File: day15/day15.hs
--import System.IO
--import Data.List (take,elemIndex)
import qualified Data.IntMap.Strict as M
--import Data.Maybe (fromJust)
 
play :: (M.Key, M.Key, M.IntMap M.Key) -> (M.Key, M.Key, M.IntMap M.Key)
play (lastspoken,lastturn,spoken) = case M.lookup lastspoken spoken of 
                                    Just y  -> (lastturn-y,lastturn+1,M.insert lastspoken lastturn spoken)
                                    Nothing -> (0         ,lastturn+1,M.insert lastspoken lastturn spoken)

geti i (lastspoken, lastturn, spoken)
    | lastturn==i = lastspoken
    | otherwise   = geti i $ play (lastspoken, lastturn, spoken)


main = do
    let spoken = M.fromList $ zip [0,14,1,3,7] [1..]
    let start = (9,6, spoken) 
    print spoken
    print $ geti 2020 start
    print $ geti 30000000 start

    

File: day15/day15.cpp

#include <iostream>
#include <map>

int play(int playTo,std::map<int,int> spoken, int lastspoken, int lastturn) {
    int next=0;
    while (lastturn<playTo) {
        if (spoken.contains(lastspoken)) {
            next=lastturn-spoken[lastspoken];
        } else {
            next=0;
        }
        spoken.insert_or_assign(lastspoken,lastturn);
        lastturn++;
        lastspoken=next;
    }
    return(lastspoken);
}


int main()
{
    std::map<int,int> spoken;

    spoken.insert({0,1});
    spoken.insert({1,3});
    spoken.insert({3,4});
    spoken.insert({7,5});
    spoken.insert({14,2});
    
    std::cout << play(30000000,spoken,9,6) << '\n';
}